gimbal-adapter/src/main/java/com/urbanairship/gimbal/GimbalAdapter.java

```diff
@@ -252,15 +248,49 @@ public void onResult(boolean enabled) {
             }
         });
 
-        task.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, ACCESS_FINE_LOCATION);
+        requestPermissionsTask.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, ACCESS_FINE_LOCATION);
+    }
+
+
+    private void startAdapter(@NonNull String gimbalApiKey) {
+        if (isAdapterStarted) {
+            return;
+        }
+
+        preferences.edit()
+                .putString(API_KEY_PREFERENCE, gimbalApiKey)
+                .putBoolean(STARTED_REFERENCE, true)
+                .apply();
+
+        Gimbal.setApiKey((Application) context.getApplicationContext(), gimbalApiKey);
+        PlaceManager.getInstance().addListener(placeEventListener);
+        updateDeviceAttributes();
+
+        Gimbal.setApiKey((Application) context.getApplicationContext(), gimbalApiKey);
```

Comment:
Api Key was already set previously. 
Is it really necessary to set the Api Key twice, add the listener twice and update device attributes twice?
I think we can remove lines 265-267

Reply:
This will only run once, its guarded by the new `isAdapterStarted` variable. However if you call stop and then start again it will do it all again.

Reply:
This has nothing to do with the variable. I think you misunderstood the comment, I wanted to point out the Code duplication inside the method. Look at lines 265-267 and then look at lines 269-272. 
Lines 269-272 look like they were duplicated and `Gimbal.start()` was added in.

Reply:
Oh, my bad. Fixed. Thanks for the great review!